### ISSUE TITLE
Name collision in extract_objects()

### ISSUE THREAD

**Issue by byron-hawkins** [mesonbuild/meson#2495]
In a c++ project, if two source files in different directories have the same name, the `extract_objects()` function will only be able to reach the first-compiled among the corresponding object files. For example:

`src/meson.build`:

    subdir('foo')

    src_files = files('Utils.cpp')
    app = executable('App', src_files, foo_files)

    subdir('test')

`src/foo/meson.build`:

    foo_files = files('Utils.cpp')

`src/test/meson.build`: 

    utils_object_1 = app.extract_objects('Utils.cpp') # extracts the Utils.o of src/Utils.cpp
    utils_object_2 = app.extract_objects(???) # can't reach Utils.o of src/util/Utils.cpp
    executable('SomeTest', 'SomeTest.cpp', objects : utils_object)

It's also confusing to reference the name `'Utils.cpp'` in a directory where there is no such file, which is currently required by the `extract_objects()` function.

**Comment by liugang** [comment]
Could you commit a simple test case, so we can checkout and do some trace quickly.

**Comment by nirbheek** [comment]
Yes, this is a known bug and it's not hard to fix; the object file naming needs to be a bit smarter.

**Comment by bonzini** [comment]
This is fixed now thanks to various changes to `object_filename_from_source` and `canonicalize_filename` (for example, QEMU uses `extract_objects` with both hw/acpi/pci.c and hw/ide/pci.c).
